mitodl/mit-open
# allow configuring Axios defaults.withCredentials

### What are the relevant tickets?
Closes https://github.com/mitodl/mit-open/issues/853

### Description (What does it do?)
This PR adds a new env setting, `MITOPEN_AXIOS_WITH_CREDENTIALS`, a boolean setting that defaults to False. This setting is passed on to webpack in the `APP_SETTINGS` global variable as `axios_with_credentials` into the `api` frontend. It is then determined in `axios.ts` if the value was true or false as a boolean, and this is set on `instance.defaults.withCredentials`. The `withCredentials` setting tells the Axios API to generate requests with `credentials: "include",` in the fetch commands that it uses. This tells the user's browser before it makes a fetch request to include any secure browser cookies that may exist for the current domain. This is useful when working with CORS requests, as credentials can be passed securely for another site you are already logged into.

### How can this be tested?
Testing the actual inclusion of cross-domain cookies is out of the scope of this PR, but you can verify that the setting is being turned on and off in your browser.

 - Set `MITOPEN_AXIOS_WITH_CREDENTIALS=True` in your `.env` file
 - Spin up `mit-open` with `docker compose up`
 - Hit http://localhost:8063/search?topic=Artificial+Intelligence&resource_type=course
 - Open your sources tab and find `axios.ts`
 - Set a breakpoint on `instance.defaults.withCredentials = withCredentials || false`
 - Refresh the page (you may need to force refresh)
 - You should see that the value of `withCredentials` is `true`
 - Set `MITOPEN_AXIOS_WITH_CREDENTIALS=True` in your `.env` file and restart your containers, repeating the process
 - This time, `withCredentials` should be `false`

### Additional Context
This functionality is being added in pursuit of the future addition of a separate webpack module bundle of exported `mit-open` components for use in other applications such as OCW
